### PR TITLE
Aarch64 codegen quality: support more general add+extend address computations.

### PR DESCRIPTION
**Pull request by cfallin**
Previously, our pattern-matching for generating load/store addresses was
somewhat limited. For example, it could not use a register-extend
address mode to handle the following CLIF:

```
   v2760 = uextend.i64 v985
   v2761 = load.i64 notrap aligned readonly v1
   v1018 = iadd v2761, v2760
   store v1017, v1018
```

This PR adds more general support for address expressions made up of
additions and extensions. In particular, it pattern-matches a tree of
64-bit `iadd`s, optionally with `uextend`/`sextend` from 32-bit values
at the leaves, to collect the list of all addends that form the address.
It also collects all offsets at leaves, combining them.
It applies a series of heuristics to make the best use of the
available addressing modes, filling the load/store itself with as many
64-bit registers, zero/sign-extended 32-bit registers, and/or an offset,
then computing the rest with add instructions as necessary. It attempts
to make use of immediate forms (add-immediate or subtract-immediate)
whenever possible, and also uses the built-in extend operators on add
instructions when possible. There are certainly cases where this is not
optimal (i.e., does not generate the strictly shortest sequence of
instructions), but it should be good enough for most code.

Using `perf stat` to measure instruction count (runtime only, on
wasmtime, after populating the cache to avoid measuring compilation),
this impacts `bz2` as follows:

```
pre:

       1006.410425      task-clock (msec)         #    1.000 CPUs utilized
               113      context-switches          #    0.112 K/sec
                 1      cpu-migrations            #    0.001 K/sec
             5,036      page-faults               #    0.005 M/sec
     3,221,547,476      cycles                    #    3.201 GHz
     4,000,670,104      instructions              #    1.24  insn per cycle
   <not supported>      branches
        27,958,613      branch-misses

       1.006071348 seconds time elapsed

post:

        963.499525      task-clock (msec)         #    0.997 CPUs utilized
               117      context-switches          #    0.121 K/sec
                 0      cpu-migrations            #    0.000 K/sec
             5,081      page-faults               #    0.005 M/sec
     3,039,687,673      cycles                    #    3.155 GHz
     3,837,761,690      instructions              #    1.26  insn per cycle
   <not supported>      branches
        28,254,585      branch-misses

       0.966072682 seconds time elapsed
```

In other words, this reduces instruction count by 4.1% on `bz2`.

### CODE REVIEW

File: cranelift/codegen/src/isa/aarch64/lower.rs [hunk 1]
```diff
@@ -544,105 +544,241 @@ pub(crate) fn alu_inst_immshift(
...
+///
+/// - NarrowValueMode::None: the associated input is 64 bits wide; no extend.
+/// - NarrowValueMode::SignExtend64: the associated input is 32 bits wide;
+///                                  do a sign-extension.
+/// - NarrowValueMode::ZeroExtend64: the associated input is 32 bits wide;
+///                                  do a zero-extension.
+///
+/// We do not descend further into the inputs of extensions, because supporting
+/// (e.g.) a 32-bit add that is later extended would require additional masking
+/// of high-order bits, which is too complex. So, in essence, we descend any
```
Comment (julian-seward1) [inline cranelift/codegen/src/isa/aarch64/lower.rs:572]: Yes.  That sounds reasonable.

File: cranelift/codegen/src/isa/aarch64/lower.rs [hunk 2]
```diff
@@ -544,105 +544,241 @@ pub(crate) fn alu_inst_immshift(
...
+    } else if offset > 0 && offset < 0x1000 && addends64.len() > 0 {
+        let reg64 = addends64.pop().unwrap();
+        let off = offset;
+        offset = 0;
+        MemArg::RegOffset(reg64, off, elem_ty)
+    } else if addends64.len() >= 2 {
+        let reg1 = addends64.pop().unwrap();
+        let reg2 = addends64.pop().unwrap();
+        MemArg::RegReg(reg1, reg2)
+    } else if addends64.len() == 0 && addends32.len() > 0 {
```
Comment (julian-seward1) [inline cranelift/codegen/src/isa/aarch64/lower.rs:670]: Can this case, `addends64.len() == 0 && <whatever>` actually happen?  Wouldn't it imply that whatever created this code knew that it would run in an environment where the max virtual address is 2^32 or some small multiple thereof ?
I guess that's implausible for the general 64-bit target case, but it's not *per se* invalid.

File: cranelift/codegen/src/isa/aarch64/lower.rs [hunk 3]
```diff
@@ -544,105 +544,241 @@ pub(crate) fn alu_inst_immshift(
...
+            from_bits: 32,
+            to_bits: 64,
+        });
+        if let Some((reg2, extendop)) = addends32.pop() {
+            MemArg::RegExtended(tmp.to_reg(), reg2, extendop)
+        } else {
+            MemArg::reg(tmp.to_reg())
         }
-    }
+    } else if addends64.len() == 0 && addends32.len() == 0 {
```
Comment (julian-seward1) [inline cranelift/codegen/src/isa/aarch64/lower.rs:690]: And again here -- even more so.  The VA is formed only from an immediate ?
Again, I guess that's OK.  But -- at least -- do we have tests for this?

File: cranelift/codegen/src/isa/aarch64/lower.rs [hunk 4]
```diff
@@ -544,105 +544,241 @@ pub(crate) fn alu_inst_immshift(
...
+        // If we can fit offset or -offset in an imm12, use an add-imm
+        // to combine the reg and offset. Otherwise, load value first then add.
+        if let Some(imm12) = Imm12::maybe_from_u64(offset as u64) {
+            ctx.emit(Inst::AluRRImm12 {
+                alu_op: ALUOp::Add64,
+                rd: addr,
+                rn: reg,
+                imm12,
+            });
+        } else if let Some(imm12) = Imm12::maybe_from_u64((-offset) as u64) {
```
Comment (julian-seward1) [inline cranelift/codegen/src/isa/aarch64/lower.rs:732]: wrapping neg needed here?

Comment (cfallin) [reply cranelift/codegen/src/isa/aarch64/lower.rs]: Done!

**Comment by cfallin** [comment]
Thanks for the review! A few answers below:

>     * the weird cases (addends64.len() == 0, etc) that I comment on -- are they all tested?

Indeed, each of the if-branches has at least one test case in `filetests/vcode/aarch64/amodes.clif`.

>     * [per recent discussion[ the CLIF adds and/or shifts that get rolled into some other instruction -- is there some mechanism that ensures that they won't be re-evaluated later?  That would be wasteful.

There isn't, but in this case I don't think it actually adds any extra work: the only ops that we bypass are `uextend`/`sextend`, and these use the free register-extend mode for the second operand in an `add`. The use-counting still applies, so if the address mode expression was the only place where the extended value was used, then the original extend op will be skipped.
 
>     * the search back up the DAG for addends -- is there some mechanism that ensures it stays in the same "color" as the root CLIF node (the load or store) at which the search is rooted?

No on this as well, but I think we are still correct: the color-crossing prohibition only applies to impure (side-effecting) ops; a sign/zero-extend is a pure function, as is an addition, so whether we keep the original source values live from many side effects ago and compute the final add/extend at use here, or pre-compute some part of the address in the distant past and carry the 64-bit value across the oceans to the present, the result should be the same. (The inputs here are still SSA, so there can't be redefinitions.) It's possible I'm missing some subtlety here though?

File: cranelift/codegen/src/isa/aarch64/lower.rs (hunk 3 above)
Comment (cfallin) [reply cranelift/codegen/src/isa/aarch64/lower.rs:691]: Yup, there's a test (`f10` in `amodes.clif`). It's conceivable that at least in a JIT scenario, one might want to bake an address into the compiled code (e.g., access a global flag somewhere). In any case, it's valid CLIF, so we should support it, I think.

File: cranelift/codegen/src/isa/aarch64/lower.rs (hunk 2 above)
Comment (cfallin) [reply cranelift/codegen/src/isa/aarch64/lower.rs:671]: Indeed, it's valid CLIF, so we need to either handle it here, or else define it as invalid and punt this problem to the CLIF verifier. It's conceivable that (as you say) the implementer might know better and use 32-bit pointers, so I don't think we can rule it out.

(I actually asserted that this case did not happen at first, but caught myself with a test-case I had previously written that used only a zero-extended 32-bit value as an address... so if nothing else, at least mischevous unit-test authors are known to do this.)

**Comment by cfallin** [comment]
Err, to caveat the above "no extra work", I realized immediately after posting that we *can* generate extra work by redoing adds. It intuitively seems unlikely to me that this would generate much extra work (emprically, on our one `bz2` test-case at least, we reduce work overall). I'm happy to tweak the logic to avoid looking through adds in other color-zones, if you'd prefer (or save that for when we update the machine-independent lowering logic to have proper rules around this).

**Comment by cfallin** [comment]
I just did a quick experiment to quantify what pulling pure ops across colors (side-effects) gives us. On 'bz2', with merging of ops only within the same color, I see:

```
       1182.421475      task-clock (msec)         #    1.000 CPUs utilized
               142      context-switches          #    0.120 K/sec
                 0      cpu-migrations            #    0.000 K/sec
             5,159      page-faults               #    0.004 M/sec
     3,630,298,549      cycles                    #    3.070 GHz
     4,201,675,603      instructions              #    1.16  insn per cycle
   <not supported>      branches
        25,751,269      branch-misses

       1.182215583 seconds time elapsed
```

Compared to the baseline (same config as post-results in this PR description):

```
        961.162550      task-clock (msec)         #    0.998 CPUs utilized          
               113      context-switches          #    0.118 K/sec                  
                 0      cpu-migrations            #    0.000 K/sec                  
             5,097      page-faults               #    0.005 M/sec                  
     3,035,892,640      cycles                    #    3.159 GHz                    
     3,837,092,873      instructions              #    1.26  insn per cycle         
   <not supported>      branches                                                    
        28,753,633      branch-misses                                               

       0.962659380 seconds time elapsed
```

So that's about a 9.5% instruction-count regression, or 22% runtime regression, if we disallow such merging. I vote we keep it :-)

**Comment by julian-seward1** [comment]
Thanks for the comments and measurements.  Sure, let's retain pulling of pure ops across colours :-)

The potential for duplicate work as a result of not forcing used-more-than-once values into a register, I'm still not happy about.  We should measure that at some point.  However, it's an issue which is far broader than this PR.  Is it easy to modify the generic lowering-ctx machinery to make that impossible, without having to change the target-dependent isel code itself?  I volunteer to do the measurements if a simple patch became available :-)

In any case, r+; please land as-is.  Thanks for chasing this around.

File: cranelift/codegen/src/isa/aarch64/lower.rs [hunk 5]
```diff
@@ -544,105 +544,242 @@ pub(crate) fn alu_inst_immshift(
...
+    roots: &[InsnInput],
+) -> (AddressAddend64List, AddressAddend32List, i64) {
+    let mut result32: AddressAddend32List = SmallVec::new();
+    let mut result64: AddressAddend64List = SmallVec::new();
+    let mut offset: i64 = 0;
+
+    let mut workqueue: SmallVec<[InsnInput; 4]> = roots.iter().cloned().collect();
+
+    while !workqueue.is_empty() {
+        let input = workqueue.pop().unwrap();
```
Comment (bnjbvr) [inline cranelift/codegen/src/isa/aarch64/lower.rs:586]: nit: `while let Some(input) = workqueue.pop()`

File: cranelift/codegen/src/isa/aarch64/lower.rs [hunk 6]
```diff
@@ -544,105 +544,242 @@ pub(crate) fn alu_inst_immshift(
...
+                    for input in 0..ctx.num_inputs(insn) {
+                        let addend = InsnInput { insn, input };
+                        workqueue.push(addend);
+                    }
+                }
+                Opcode::Iconst => {
+                    let value: i64 = ctx.get_constant(insn).unwrap() as i64;
+                    offset += value;
+                }
+                _ => {
```
Comment (bnjbvr) [inline cranelift/codegen/src/isa/aarch64/lower.rs:619]: nit: i thought this was a bug at first, because only the 4 opcodes mentioned here should be caught, and then saw the extra condition on Uextend/Sextend. Instead, would it make sense to split this match into two:

- one for Uextend and Sextend that don't match the condition, so just `Opcode::Uextend | Opcode::Sextend` after the first arm
- one `_ => panic!("unexpected opcode")`

File: cranelift/codegen/src/isa/aarch64/lower.rs [hunk 7]
```diff
@@ -544,105 +544,242 @@ pub(crate) fn alu_inst_immshift(
...
     elem_ty: Type,
-    addends: &[InsnInput],
+    roots: &[InsnInput],
     offset: i32,
 ) -> MemArg {
-    // TODO: support base_reg + scale * index_reg. For this, we would need to pattern-match shl or
-    // mul instructions (Load/StoreComplex don't include scale factors).
+    // TODO: support base_reg + scale * index_reg. For this, we would need to
+    // pattern-match shl or mul instructions (Load/StoreComplex don't include
+    // scale factors).
```
Comment (bnjbvr) [inline cranelift/codegen/src/isa/aarch64/lower.rs:642]: Fyi, the comment line size is 99 chars, so the previous wrapping seemed somewhat correct.

File: cranelift/codegen/src/isa/aarch64/lower.rs [hunk 8]
```diff
@@ -544,105 +544,242 @@ pub(crate) fn alu_inst_immshift(
...
-                parts = Some((
-                    r1,
-                    r2,
-                    r2_bits,
-                    /* is_signed = */ op == Opcode::Sextend,
-                ));
-                break;
-            }
-        }
+    debug!(
```
Comment (bnjbvr) [inline cranelift/codegen/src/isa/aarch64/lower.rs:650]: maybe trace?

File: cranelift/codegen/src/isa/aarch64/lower.rs [hunk 9]
```diff
@@ -544,105 +544,242 @@ pub(crate) fn alu_inst_immshift(
...
+            MemArg::reg(tmp.to_reg())
         }
-    }
+    } else if addends64.len() == 0 && addends32.len() == 0 {
+        let off_reg = ctx.alloc_tmp(RegClass::I64, I64);
+        lower_constant_u64(ctx, off_reg, offset as u64);
+        offset = 0;
+        MemArg::reg(off_reg.to_reg())
+    } else {
+        let reg1 = addends64.pop().unwrap();
```
Comment (bnjbvr) [inline cranelift/codegen/src/isa/aarch64/lower.rs:697]: From just skimming this code, it's hard to believe that this last branch is hit when `addends64.len() > 0`. Could the whole series be reorganized as `if addends64.len() > 0 { /* subchecks, etc. */ } else /*other subchecks*/{}`?

File: cranelift/codegen/src/isa/aarch64/lower.rs (hunk 7 above)
Comment (cfallin) [reply cranelift/codegen/src/isa/aarch64/lower.rs]: Ah, I had my editor set to 80 characters and probably just rewrapped this while nearby. Undid change, thanks.

File: cranelift/codegen/src/isa/aarch64/lower.rs (hunk 6 above)
Comment (cfallin) [reply cranelift/codegen/src/isa/aarch64/lower.rs]: Good idea, thanks!

File: cranelift/codegen/src/isa/aarch64/lower.rs (hunk 8 above)
Comment (cfallin) [reply cranelift/codegen/src/isa/aarch64/lower.rs]: Done.

File: cranelift/codegen/src/isa/aarch64/lower.rs (hunk 9 above)
Comment (cfallin) [reply cranelift/codegen/src/isa/aarch64/lower.rs]: Ah yes, it's much cleaner with that split -- thanks!

File: cranelift/codegen/src/isa/aarch64/lower.rs (hunk 5 above)
Comment (cfallin) [reply cranelift/codegen/src/isa/aarch64/lower.rs]: Done.

**Comment by cfallin** [comment]
@julian-seward1, to close the loop on the merging-rules discussion, I wrote another quick patch to limit merges to only instructions with one use, available as the latest commit here (no merge, may break, in particular `ifcmp` with multiple uses not merging would break things, but `bz2` seems OK without this): [commit in branch eval-merging-restrictions](https://github.com/cfallin/wasmtime/commit/ba0d275b171b7364fb8176fe1103da76a23d2871)

With the patch applied, which limits merging to 1-use-only instructions as well as same-color (so it includes the above restriction too), on `bz2` I see a 16.8% regression in instruction count and a 24.8% regression in runtime. This matches my intuition that at least on aarch64, pulling pure ops (extends, etc) into instructions is pretty important. Please do poke holes in the patch and/or play with it more, though, if desired!
